### ISSUE TITLE
Race condition: replies, deletes, undos, unfollows, etc get dropped if their target post is still processing

### ISSUE THREAD

**Issue by qazmlp** [snarfed/bridgy-fed#1361]
A reply just (sporadically?) didn't make it to Bluesky. The log at https://fed.brid.gy/log?path=%2Fqueue%2Fwebmention%2C%2Finbox&start_time=1728105708&key=https%3A%2F%2Fwolfdo.gg%2Fnotes%2F9yz9tw8dfgws6ex9%2Factivity shows the following

```
I 2024-10-05 05:21:46.205812+00:00 Already seen this activity https://wolfdo.gg/notes/9yz9tw8dfgws6ex9/activity
```

Relevant thread on Bluesky: https://bsky.app/profile/qazm.bsky.social/post/3l5qmbrxis223

That looks like a race condition somewhere in the ActivityPub inbox to me.
(This activity doesn't have the green checkmark for delivering it to Bluesky.)

**Comment by snarfed** [comment]
Hah, race condition indeed. That reply (call it B) was to another reply (A) by the same author that was still processing when B came in. BF saw that A wasn't bridged, so it didn't bridge B either, due to the logic that replies to non-bridged posts also aren't bridged.

Ugh. I can think of ways to solve this, but they're all complicated. I don't see an easy fix yet.

(The log message was unrelated, we got two inbox deliveries of this reply and that log message was for the second one. Our logic for finding log messages is not the best 🤷)

**Comment by TomCasavant** [comment]
Let me know if you want me to file a different issue for this, but I'm pretty sure this is related to this issue

https://fed.brid.gy/ap/@surrender_idx90@tomkahe.com <- this bot auto reposts any post from https://fed.brid.gy/ap/@surrender_index@tomkahe.com if certain conditions are met and it boosts them basically immediately after the post is created. It looks like Bridgy is seeing the boost but doesn't do anything about it (I'm guessing because Bridgy hasn't processed the original post yet)

**Comment by snarfed** [comment]
Ah, interesting, true! That does sound likely.

**Comment by snarfed** [comment]
Got another report of this, #1393 ^, for deletes. Probably the same issue.

One way to fix this could be to error out if we get a reply, repost, quote, delete, or undo of an object that either we haven't seen yet or isn't bridged. We'd then retry the receive task 5m later, and then again 10m after that.

I'm reluctant to do that though, since we get a lot of these, and the _vast_ majority of the time, the object in question will never be bridged. We'd end up tripling our load for those incoming activities to fix a rare edge case. 😕

**Comment by TomCasavant** [comment]
FWIW, I ended up fixing *my* issue by just delaying the boost by 30 seconds. Though I'm not sure how useful that info is for fixing the actual issue

**Comment by qazmlp** [comment]
One complication is that I don't think there's any guarantee an e.g. Create won't arrive after its matching Delete in ActivityPub (unless the instance cancels relevant outbox items (synchronously), which Mastodon doesn't appear to do at all).

Backoff due to connectivity issues can sporadically (and individually explicitly [somewhat randomly](https://github.com/mastodon/mastodon/blob/1fc165de02d79294c8a218f5fa82bcd477484ca1/app/workers/activitypub/delivery_worker.rb#L19)!) reorder activities across a long time span, at least in Mastodon.

Mastodon fixes this problem with a database lock on the post id and by tombstoning e.g. statuses it receives a Delete for.

**Comment by snarfed** [comment]
Related: #1338

**Comment by atouu** [comment]
> Got another report of this, #1393 ^, for deletes. Probably the same issue.
> 
> One way to fix this could be to error out if we get a reply, repost, quote, delete, or undo of an object that either we haven't seen yet or isn't bridged. We'd then retry the receive task 5m later, and then again 10m after that.
> 
> I'm reluctant to do that though, since we get a lot of these, and the _vast_ majority of the time, the object in question will never be bridged. We'd end up tripling our load for those incoming activities to fix a rare edge case. 😕

This did happened to me too! I forgot to mark my post as mention only lol, and deleted it second after it get posted.

Bsky post link: https://bsky.app/profile/mashi.wip.la/post/3laciq5hih5p2
Masto post link: https://masto.ai/@atou/113437693154993234

**Comment by dos1** [comment]
Just stumbled upon it as well :laughing: https://bsky.app/profile/dos.social.librem.one.ap.brid.gy/post/3lau4ps7fqe62

It seems quite serious to me. If I delete something soon after posting, it's very likely because I really don't want it to stay there.

**Comment by Donnnno** [comment]
Ah, this happened to me too! :')
https://bsky.app/profile/neppy.cat/post/3lb43ta3tbyl2
https://bsky.app/profile/neppy.cat/post/3lb7vejx4gfw2

**Comment by atouu** [comment]
Same, it'll be nice if there's at least a way to delete it

**Comment by Tamschi** [comment]
I should probably put the `safety` label on this too, since it can lead to situations where the user is unable to depublish a post (aside from taking the whole account offline).

@snarfed I think there does need to be some level of general `Delete` persistence, at least due to https://github.com/snarfed/bridgy-fed/issues/1361#issuecomment-2425172677.

One option would be to have any `Delete`s by at least previously bridged accounts tombstone-flag posts in the database by their URI. That could also act as a synchronisation point, where the sending task first creates a pending-flag in that same place, which, if the `Delete` handler sees that while placing the tombstone, causes the latter to optimistically delete the post in the PDS.

If the `Delete`-handler was too fast, then the `Create` or `Update` handler will see the tombstone when it checks after its other tasks and optimistically delete the post in the PDS on its end. Individually each delete towards ATProto could fail, but overall at least one of them should go through.

That should be fairly reasonable load-wise, at least if the URIs can be prefix-deduplicated or otherwise compressed in the database.

**Comment by snarfed** [comment]
Agreed on the safety label and priority here. Re design, we could definitely add persistence for `Delete`s, but I'm still inclined to try https://github.com/snarfed/bridgy-fed/issues/1361#issuecomment-2424248716 first, ie just retry `Delete`s and `Undo`s where we don't have a matching object or activity yet. Simpler, and with our current config of retrying after 5m and then again 10m later, we'd have a 15m total window to wait for the missing object/activity. Should almost always be enough!

**Comment by snarfed** [comment]
Another approach here could be to delay initial processing of `Delete`s and `Undo`s by a minute or two. That should catch the vast majority of these, enough that we might not need to add the retries and load I worried about.

**Comment by Archenoth** [comment]
One possibly-easier way to work around this in the short term could be to add a button to posts on `https://fed.brid.gy/ap/<account>` pages to "refresh" individual posts

![image](https://github.com/user-attachments/assets/27ff5f4f-b027-491e-b275-b21c879f09bc)

That way, if someone notices weirdness, they could click something to fix it!

Also, because this would just be a "refresh", you also wouldn't need to worry about authentication

**Comment by snarfed** [comment]
^ I shipped a 2m delay for delete, undo, and stop-following, which should hopefully fix this. Fingers crossed!

**Comment by atouu** [comment]
Thank you! But how about the existing deletes that failed to bridge before? Is there anything we can do?

**Comment by snarfed** [comment]
@atouu I can delete them manually. I've already deleted the one you mentioned in https://github.com/snarfed/bridgy-fed/issues/1361#issuecomment-2466634080 . Let me know if you have others!

**Comment by Archenoth** [comment]
Like, in this issue directly? Or..?
(Because I imagine there are quite a few of these between the people who might be reading this)

**Comment by Daft-Freak** [comment]
Sooo... this issue was originally about replies, which still have the problem. As an example https://fosstodon.org/@Daft_Freak/113600878630847815 vs https://bsky.app/profile/charlie.daft.games/post/3lckxmx24viz2. The first reply (two minutes later) didn't make it through.

**Comment by snarfed** [comment]
Argh, great point! Thank you very true. Reopening.

I'm reluctant to add the same two-minute delay to replies that I added to deletes., but I don't have a better solution yet. Hrm.

**Comment by Tamschi** [comment]
> Argh, great point! Thank you very true. Reopening.
> 
> I'm reluctant to add the same two-minute delay to replies that I added to deletes., but I don't have a better solution yet. Hrm.

That wouldn't fix it for replies to replies.

**Comment by snarfed** [comment]
The other form of this problem is when we get two dependent activities in the _right_ order, but the first one isn't done before the second one that depends on it starts. Example: https://bsky.app/profile/mackuba.eu/post/3ltuxyoute222

For example, when we get a reply or like or repost, we first check that the original post is bridged. If it's not, we do nothing. However, we might be in the middle of bridging it at that moment, and we're just not done yet.

One approach here would be to generalize the current 2m delay for deletes. When we see a dependent activity, and its dependency isn't bridged but could be, delay and retry a few times, with increasing backoff.

**Comment by snarfed** [comment]
Phew, took some work, but I finally landed a fix here that I'm ok with and have some confidence in, c7f538f. Basically the design in https://github.com/snarfed/bridgy-fed/issues/1361#issuecomment-3067431977 : when we get an activity that depends on another activity that's not yet bridged, but could be, we stop processing it and retry 30s later, and repeat, with exponential backoff, up to 5 attempts total.

Not easy to test, but I tried, and I think it's working.
